### PR TITLE
Implement HTTP Response status code metrics reporting for REST API

### PR DESCRIPTION
**Pull request by miguelsorianod**
Hi,

As part of https://issues.redhat.com/browse/IPT-109, https://issues.redhat.com/browse/IPT-155 I'm trying to come up with a solution on reporting metrics with the response codes status and durations (latencies) in the REST API.

I've taken a look at the codebase to see how API rest is exposed and this is how I think it works:

It seems that metrics in service-registry are reported/exposed using the `Microprofile` (https://projects.eclipse.org/projects/technology.microprofile) Java library. Metrics are exposed on the `/metrics` endpoint.
The REST Web Services API is exposed using `JAX-RS` (https://docs.oracle.com/javaee/7/api/javax/ws/rs/package-summary.html) . The REST API is exposed in the `/api` endpoint.

Based on this I explored the following approaches:
* Trying to use some possible existing construct in `Microprofile` that reports HTTP metrics. I haven't found any construct to do that. Keep in mind I have never used `Microprofile` before. Maybe there is something but if there is I haven't found it.
*  It seems using a standard Java interceptor like the one we have for rest metrics `RestMetricsInterceptor` cannot be used as standard interceptors do not receive detail about HTTP responses

After digging a little bit on what could be done a solution that I think might be feasible to implement is making use of `ContainerResponseFilter` in JAX-RS (https://docs.oracle.com/javaee/7/api/javax/ws/rs/container/ContainerResponseFilter.html)  that seems allows us to filter HTTP Responses.

This PR implements reporting 1xx, 2xx, 3xx, 4xx, 5xx response codes counts on requests performed to the REST API. It performs a NameBinding of the ContainerResponseFilter with the `RegistryApplication` class that I understand contains the entrypoint of the REST API. This means for example that a request to /anotherurl would not be processed by this filter.

There are some specificities of the implemented behavior that we should evaluate whether they are in our interest or not:

I've done a few tests to see when the filter is executed with the current implementation:
* When a 200 status code is returned the response filter is executed
* When a 503 status code is returned the response filter is executed
* When a 406 status code is returned (forced for example by sending '-H "Accept: text/xmsdfdsfl') the response filter is NOT executed. This behavior surprised me
* When a 400 status code is returned (forced for example by sending invalid data like -d 'invaliddata') the response filter is executed
* When a 404 status code is returned due to a non-existing endpoint has been requested the response filter is NOT executed. I think this behavior is expected as
  we are only filtering on the existing api rest endpoints. In any case, is this what we want?
  endpoints
* When executing a request against /metrics for example, or to /nonexistentpath (which are not inside the api rest code) the response filter is
  NOT executed. I think this behavior is expected as they are not prat of the REST api code. In any case, is this what we want?

I've not tried with other kind of status codes, but it seems that there's some case/s where a request to an existing endpoint is not
evaluated by the filter for some reason.

The implemented metrics format look like this:
```
msoriano@localhost:~/eclipse-workspace/github.com/Apicurio/apicurio-registry (add-rest-api-metrics-response-filter)$ curl localhost:8080/metrics 2>/dev/null| grep -i xx
application_rest_response_status_code_total{code="4xx",group="REST",metric="rest_response_status_code"} 2.0
application_rest_response_status_code_total{code="5xx",group="REST",metric="rest_response_status_code"} 3.0
```

The corresponding metric is created the first time a new status code family type is returned.

We could add more tags/labels to the metric or organize different metrics for it. Just keep in mind that depending on what we process and the defined metrics
the cardinality of the metrics could increase expontentially so we have to be careful with it, specially if we add a label/metric that might have unbounded values.

What do you think about this implementation approach and the current behavior? let's iterate on this.

Thank you.

### CODE REVIEW

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+	private static final Logger log = LoggerFactory.getLogger(RestMetricsResponseFilter.class);
+
+	String REST_RESPONSE_STATUS_CODE_COUNT = "rest_response_status_code";
+	String REST_RESPONSE_STATUS_CODE_COUNT_DESC = "Total number of REST HTTP Response Codes across all endpoints.";
+
+	@Override
+	public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext)
+			throws IOException {
+		log.info("'{} {}'", requestContext.getMethod(), requestContext.getUriInfo().getRequestUri().toString());
```
Comment (miguelsorianod) [inline app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java:40]: This is info now but I think we should change it to debug or just removing it. Leaving it here for the moment until more iteration is done to the PR

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import org.eclipse.microprofile.metrics.Metadata;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.eclipse.microprofile.metrics.Tag;
+import org.eclipse.microprofile.metrics.annotation.RegistryType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Provider
+@RestMetricsResponseFilteredNameBinding
+public class RestMetricsResponseFilter implements ContainerResponseFilter {
```
Comment (carlesarnal) [inline app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java:26]: I would replace the class name by something more meaningful, maybe `RegisterStatusCodeResponseFilter`.

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilteredNameBinding.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package io.apicurio.registry.metrics;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import javax.ws.rs.NameBinding;
```
Comment (carlesarnal) [inline app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilteredNameBinding.java:8]: Same here, I would replace this name by something more related to its behaviour.

**Comment by carlesarnal** [comment]
Hi Miguel, after having a look at the PR the code looks good to me.

Regarding your comments, the only one that surprises me a bit is the one regarding the `406`. Regarding the request to non-existing paths, or paths not under `/api`, that's what I was expecting.

**Comment by miguelsorianod** [comment]
I've digged a little bit further about whether we might have something out of the box provided by Microprofile and the answer is no:

On our project we are using quarkus with the metrics extension (smallrye-metrics) and the JAX-RS rest extension (resteasy-jackson).

In resteasy-extension rest metrics can be enabled by setting the 'quarkus.resteasy.metrics.enabled=true' property in app/src/main/resources/application.properties which is what we wanted, but after enabling them the information that they provide is not enough.

See an example of what's provided:
```
+# HELP base_REST_request_total The number of invocations and total response time of this RESTful resource method since the start of the server.
+# TYPE base_REST_request_total counter
+base_REST_request_total{class="io.apicurio.registry.rest.ArtifactsResource",method="listArtifacts"} 1.0
+# TYPE base_REST_request_elapsedTime_seconds gauge
+base_REST_request_elapsedTime_seconds{class="io.apicurio.registry.rest.ArtifactsResource",method="listArtifacts"} 0.020181832
```

The format of this metrics is defined in the spec of Microprofile https://download.eclipse.org/microprofile/microprofile-metrics-2.3/microprofile-metrics-spec-2.3.html#_optional_rest that implementations like quarkus-resteasy implement.

Basically a counter and a gauge for each rest method is provided but not with information about response codes or any other kind of data.

As an additional observation, it seems that 406 errors are also not counted against this exposed metrics by the resteasy extension (when enabled).

In summary, provided metrics by resteasy seem to be not enough for our purposes so I think we have to write something ourselves as this PR is trying to do (unless I'm missing another possible solution)

**Comment by akieling** [comment]
@miguelsorianod you are in the right track. The link below shows how I did something similar in the Syndesis project. It uses Micrometer instead of microprofile metrics but should be easy to understand. It not only counts requests but also times the requests and creates a histogram of that. Basically, the same code creates the metrics required to calculate availability and latency.

https://github.com/syndesisio/syndesis/blob/master/app/server/runtime/src/main/java/io/syndesis/server/runtime/EndpointMetricsFilter.java

Based on [Prometheus naming conventions](https://prometheus.io/docs/practices/naming/) I'd suggest calling it `http_requests_total` (or something similar).

Also, unless there is a reason to include the `group` and `metric` labels, I'd suggest removing them.

**Comment by miguelsorianod** [comment]
Hi @akieling,

I think providing the path of the URL might be potentially dangerous due to possible number of elements are unbounded.
For example:
```
application_rest_http_requests_total{endpoint="/artifacts",group="REST",method="GET",metric="rest_http_requests_total",status="2xx"} 1.0
application_rest_http_requests_total{endpoint="/artifacts",group="REST",method="POST",metric="rest_http_requests_total",status="2xx"} 1.0
application_rest_http_requests_total{endpoint="/artifacts/3",group="REST",method="PUT",metric="rest_http_requests_total",status="4xx"} 1.0
application_rest_http_requests_total{endpoint="/artifacts/6",group="REST",method="PUT",metric="rest_http_requests_total",status="4xx"} 1.0
application_rest_http_requests_total{endpoint="/artifacts/mytest",group="REST",method="PUT",metric="rest_http_requests_total",status="4xx"} 1.0
```

would be two generated metrics because they match the '"/{artifactId}"' put method of the REST API and are thus processed y the filter.
This would potentially make the number of prometheus metrics increase in an unbounded way (and thus the storage and processing time needed for them)

How did you handle that? did you assume those potential unbounded results?

**Comment by akieling** [comment]
@miguelsorianod You need to set the label value like "/artifacts/{id}", not "/artifacts/1", "/artifacts/2", etc.
See https://github.com/syndesisio/syndesis/blob/3d1b52cafaa5a968fc9c83a8e36545df656539c4/app/server/runtime/src/main/java/io/syndesis/server/runtime/EndpointMetricsFilter.java#L85

**Comment by miguelsorianod** [comment]
> @miguelsorianod You need to set the label value like "/artifacts/{id}", not "/artifacts/1", "/artifacts/2", etc.
> See https://github.com/syndesisio/syndesis/blob/3d1b52cafaa5a968fc9c83a8e36545df656539c4/app/server/runtime/src/main/java/io/syndesis/server/runtime/EndpointMetricsFilter.java#L85

Hi,

The code you referenced is about the method (GET, POST, ...) and not about the URI :thinking:

**Comment by miguelsorianod** [comment]
I think I see it now,

It constructs the URI making use of the ResourceInfo class provided with context and then it tries to look for the @Path annotations at class and method level to get the information of the URL instead of obtaining it from RequestContext's getMethod method.

Now:
```
# TYPE application_rest_http_requests_total counter
application_rest_http_requests_total{endpoint="/artifacts",group="REST",method="POST",metric="rest_http_requests_total",status="2xx"} 1.0
application_rest_http_requests_total{endpoint="/artifacts/{artifactId}",group="REST",method="PUT",metric="rest_http_requests_total",status="4xx"} 2.0
```

**Comment by akieling** [comment]
My bad for the incorrect link. Yes, that's how you create your endpoint label value. Any changes to the annotation will probably change the value but that's very unlikely.

**Comment by miguelsorianod** [comment]
I've updated the PR to include HTTP method and endpoint in the count metric and renamed that metric to application_rest_http_requests_total.
Method endpoint values have the form "/{artifactId}" as proposed in previous comments so the number of values is not unbounded.

Additionally, I've updated the PR to include a duration-related metric as originally requested in https://issues.redhat.com/browse/IPT-155. It's been implemented using a Timer, which as part of the metrics that generates includes a Summary. A discussion is ongoing on the linked issue related to whether the usage of summaries is enough or not. The name of the metric is application_rest_http_request_duration

The metrics of this PR still have the 'metric' and 'group' labels. Let's see what reviewers' opinion in on whether keeping them or not.

**Comment by akieling** [comment]
@miguelsorianod Note the Timer includes a counter, therefore we might not need two separate metrics for availability and latency.

**Comment by miguelsorianod** [comment]
I saw that. I think we can and should keep the counter. The reason is they are conceptually different metrics.

Some elements include counters to perform their needed calculations, as they do with gauges too. That's just how Summaries are implemented in Prometheus. You can see that throughout metrics in all kind of products. Histograms also include counters, gauges, etc...  .

**Comment by miguelsorianod** [comment]
Hi,

With the current approach we can measure latencies (with the Timer) but we cannot aggregate them between instances, for that an histogram would be needed (see the details about this in https://issues.redhat.com/browse/IPT-155).

I think for the moment, even if it does not work in some scenarios, we could merge this so we provide something. Keeping in mind we might do some changes/addition/removal (even potentially large ones) of exposed metrics or even the metrics framework in the future depending on the solution decided to be taken in the  https://issues.redhat.com/browse/IPT-155 issue to solve the latencies-related metrics issue.

cc @EricWittmann @jsenko @akieling to review current implementation and thoughts on this.

Thank you.

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java (hunk 1 above)
Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java:64]: Agreed - should be either debug or trace level.

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java (hunk 2 above)
Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java]: Unless we think the filter is going to track additional metrics.  But if we're renaming I'd suggest `RestStatusCodeMetricsResponseFilter`.

**Comment by EricWittmann** [comment]
Hey @miguelsorianod - thanks for this contribution!  Sorry for the lateness of my review on this.  Going through the code now and will merge soon if no one objects.

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import org.eclipse.microprofile.metrics.Timer;
+import org.eclipse.microprofile.metrics.annotation.RegistryType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.smallrye.metrics.app.Clock;
+
+@Provider
+@RestMetricsResponseFilteredNameBinding
+public class RestMetricsResponseFilter implements ContainerRequestFilter, ContainerResponseFilter {
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java:47]: Can you please add a javadoc comment at the top to explain what this filter is doing?

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java (hunk 1 above)
Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java]: After thinking about this more, let's delete this. :) I think it is redundant since we have logging enabled (via annotation) for all the REST implementation classes.

Comment (miguelsorianod) [reply app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java]: :+1:

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java (hunk 2 above)
Comment (miguelsorianod) [reply app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java]: I think we should keep the current name as there is other information other than status code being gathered

File: app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java (hunk 4 above)
Comment (miguelsorianod) [reply app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java:47]: Ok, going to do it now :+1:

Comment (miguelsorianod) [reply app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java:54]: Documentation added. @EricWittmann is there something pending?

Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/metrics/RestMetricsResponseFilter.java:54]: Hey @miguelsorianod we've been working on getting the build to work properly in CI and I've been trying to recover from hurricane outages.  I haven't forgotten about this!
